### PR TITLE
WIP Factor-out auth-related logic

### PR DESCRIPTION
**Pull request by nl0**
Factor-out auth-related logic and adjust it a little. Log-out on auth failure.

Depends on:
  - [x] #589 
  - [x] #590 
  - [x] #591 
  - [x] #596 
  - [x] #597 
  - [x] #599 

### TODO

- [x] Unit Tests
- [x] rebase
- [ ] more manual testing
- [x] figure out 500 vs 401 behaviour
- [x] styling
- [ ] rewrite tests

### CODE REVIEW

**Comment by nl0** [comment]
@meffij 
> Does this log the user out if a request times out?

yes

> A 500 or a timeout should do exponential backoff retry

makes sense, but this will add substantial amount of complexity

> and a 401 should log the user out

this is the default behaviour: drop the tokens and log user out if any of auth requests (refreshing tokens or getting auth data) fails

**Comment by meffij** [comment]
If you don't want to tackle exponential backoff, then just do whatever we currently do for a 500 or timeout. I don't want users to be logged out of the catalog just because their Internet is spotty or there's some issue with the backend.

**Comment by meffij** [comment]
TODOs for the new auth work:
- [ ] `/login`
- [ ] `/logout`
- [ ] `/signup`
- [ ] `/reset_password`
- [ ] `/reset_password/$link`
- [ ] password reset link on login page -- links to `/reset_password`
- [ ] `/code`

**Comment by nl0** [comment]
@akarve @meffij so, everything seems to be working (with fixes from #633). next ima tweak some styling and rewrite the tests, and then we're good

File: catalog/app/containers/Auth/messages.js [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+    id: `${scope}.SignIn.hintSignUp`,
+    defaultMessage: "Don't have an account? {link}.",
+  },
+  signInHintSignUpLink: {
+    id: `${scope}.SignIn.hintSignUpLink`,
+    defaultMessage: 'Sign up',
+  },
+  signInHintReset: {
+    id: `${scope}.SignIn.hintReset`,
+    defaultMessage: "Don't remember your password? {link}.",
```
Comment (akarve) [inline catalog/app/containers/Auth/messages.js]: "Did you forget your password?"

File: catalog/app/containers/Auth/messages.js [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+  },
+
+  // SignUp
+  signUpHeading: {
+    id: `${scope}.SignUp.heading`,
+    defaultMessage: 'Sign Up',
+  },
+  signUpPassResetHint: {
+    id: `${scope}.SignUp.passResetHint`,
+    defaultMessage: "Dont't remember your password?",
```
Comment (akarve) [inline catalog/app/containers/Auth/messages.js]: "Did you forget your password?"

File: catalog/app/containers/Auth/tests/__snapshots__/Auth.spec.js.snap [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+`;
+
+exports[`Feature: containers/Auth/SignOut Background: Given minimal setup Scenario: Mounting the component when auth data is stale and then gets refreshed with error Given storage has stale auth data When the component tree is mounted When auth refresh requests are rejected Then the user should be signed out 1`] = `
+<div>
+  <div
+    __name="Home"
+  />
+  <div
+    __name="Notification"
+    message="Authentication lost"
```
Comment (akarve) [inline catalog/app/containers/Auth/tests/__snapshots__/Auth.spec.js.snap]: Can we have a better message? What has actually happend at this point? And should we just log people out?

File: catalog/app/containers/Auth/messages.js [hunk 4]
```diff
@@ -0,0 +1,281 @@
+import { defineMessages } from 'react-intl';
+
+const scope = 'app.containers.Auth';
+
+export default defineMessages({
+  notificationAuthLost: {
+    id: `${scope}.notificationAuthLost`,
+    defaultMessage: 'Authentication lost. Please sign in again.',
```
Comment (akarve) [inline catalog/app/containers/Auth/messages.js]: Same question, when does this happen? "~~Please~~Sign in again."

File: catalog/app/containers/Auth/messages.js [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+    id: `${scope}.SignUp.usernameLabel`,
+    defaultMessage: 'Username',
+  },
+  signUpUsernameRequired: {
+    id: `${scope}.SignUp.usernameRequired`,
+    defaultMessage: 'Enter a username',
+  },
+  signUpUsernameTaken: {
+    id: `${scope}.SignUp.usernameTaken`,
+    defaultMessage: 'This username is already taken. {link}',
```
Comment (akarve) [inline catalog/app/containers/Auth/messages.js]: 'Username taken. {link}'

File: catalog/app/containers/Auth/messages.js [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+  },
+  signUpUsernameInvalid: {
+    id: `${scope}.SignUp.usernameInvalid`,
+    // TODO: specify username requirements
+    defaultMessage: 'This username is invalid',
```
Comment (akarve) [inline catalog/app/containers/Auth/messages.js]: 'Username invalid'

Comment (akarve) [reply catalog/app/containers/Auth/messages.js]: Yeah, ask calvin for requirements. I think it's 8 characters and that's it.

File: catalog/app/containers/Auth/messages.js [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+  },
+  signUpEmailLabel: {
+    id: `${scope}.SignUp.emailLabel`,
+    defaultMessage: 'Email',
+  },
+  signUpEmailRequired: {
+    id: `${scope}.SignUp.emailRequired`,
+    defaultMessage: 'Enter your email',
```
Comment (akarve) [inline catalog/app/containers/Auth/messages.js]: Is this hint text? If so 'you@example.com' is better.

File: catalog/app/containers/Auth/messages.js [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+  },
+  signUpEmailTaken: {
+    id: `${scope}.SignUp.emailTaken`,
+    defaultMessage: 'This email is already taken. {link}',
```
Comment (akarve) [inline catalog/app/containers/Auth/messages.js]: 'Email taken...'

File: catalog/app/containers/Auth/messages.js [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+  },
+  signUpHintSignInLink: {
+    id: `${scope}.SignUp.hintSignInLink`,
+    defaultMessage: 'Sign in',
+  },
+
+  // SignIn
+  signInHeading: {
+    id: `${scope}.SignIn.heading`,
+    defaultMessage: 'Sign In',
```
Comment (akarve) [inline catalog/app/containers/Auth/messages.js]: We usually say "Sign in" (lowercase i)

File: catalog/app/containers/Auth/messages.js [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+  },
+  signInUsernameLabel: {
+    id: `${scope}.SignIn.usernameLabel`,
+    defaultMessage: 'Username',
+  },
+  signInUsernameRequired: {
+    id: `${scope}.SignIn.usernameRequired`,
+    defaultMessage: 'Enter your username',
```
Comment (akarve) [inline catalog/app/containers/Auth/messages.js]: Hint text? Maybe "your_name" is better?

File: catalog/app/containers/Auth/messages.js [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+    id: `${scope}.SignIn.passLabel`,
+    defaultMessage: 'Password',
+  },
+  signInPassRequired: {
+    id: `${scope}.SignIn.passRequired`,
+    defaultMessage: 'Enter your password',
+  },
+  signInSubmit: {
+    id: `${scope}.SignIn.submit`,
+    defaultMessage: 'Sign In',
```
Comment (akarve) [inline catalog/app/containers/Auth/messages.js]: lowercase i

File: catalog/app/containers/Auth/messages.js (hunk 6 above)
Comment (nl0) [reply catalog/app/containers/Auth/messages.js]: @meffij could you please clarify?

File: catalog/app/containers/Auth/messages.js (hunk 10 above)
Comment (nl0) [reply catalog/app/containers/Auth/messages.js]: this is an error message as well

File: catalog/app/containers/Auth/messages.js (hunk 4 above)
Comment (nl0) [reply catalog/app/containers/Auth/messages.js]: this happens if a user is authenticated, but the next auth check receives 401 from the server (for example, if user's access was revoked, or if the token was invalidated for some reason)

File: catalog/app/containers/Auth/messages.js (hunk 6 above)
Comment (meffij) [reply catalog/app/containers/Auth/messages.js]: Currently, the only restriction is that the password must be at least 8 characters long.

Comment (nl0) [reply catalog/app/containers/Auth/messages.js]: @meffij cool, but what are the requirements for username?

**Comment by nl0** [comment]
closing in favor of #657

**Pull request closed without merging**